api/common/sharing/utils.go

```diff
@@ -12,44 +12,234 @@ import (
+			return types.ServiceInstances{}, err
+		}
+	}
+
+	for i := 0; i < instances.Len(); i++ {
+
+		// "*" for any shared instance of the tenant, if parameter not passed, set true for any shared instance
+		selectorVal, exists := parameters[instance_sharing.ReferencedInstanceIDKey]
+		if exists && selectorVal.String() != "*" {
+			continue
+		}
+
+		instance := instances.ItemAt(i).(*types.ServiceInstance)
+
+		selectorVal, exists = parameters[instance_sharing.ReferenceInstanceNameSelector]
+		if exists && len(selectorVal.String()) > 0 && !(instance.Name == selectorVal.String()) {
+			continue
+		}
+
+		if planSelectorExists && (selectorPlan == nil || instance.ServicePlanID != selectorPlan.ID) {
+			continue
+		}
+
+		selectorVal, exists = parameters[instance_sharing.ReferenceLabelSelector]
+		if exists && len(selectorVal.String()) > 0 {
+			match, err := matchLabels(instance.Labels, []byte(selectorVal.Raw))
+			if err != nil {
+				return types.ServiceInstances{}, err
+			}
+			if !match {
+				continue
+			}
+		}
+
+		filteredInstances.Add(instance)
+		// only single result is accepted for selectors:
+		if filteredInstances.Len() > 1 {
+			log.C(ctx).Errorf("%s", util.ErrMultipleReferenceSelectorResults)
+			return types.ServiceInstances{}, util.HandleInstanceSharingError(util.ErrMultipleReferenceSelectorResults, "")
+		}
+	}
+	if filteredInstances.Len() == 0 {
+		log.C(ctx).Errorf("%s", util.ErrNoResultsForReferenceSelector)
+		return types.ServiceInstances{}, util.HandleInstanceSharingError(util.ErrNoResultsForReferenceSelector, "")
+	}
+	return filteredInstances, nil
+}
+
+func getSelectorPlan(ctx context.Context, repository storage.Repository, smaap bool, offeringID, selectorValue string) (*types.ServicePlan, error) {
+	var err error
+	var key string
+	if smaap {
+		key = "name"
+	} else {
+		key = "catalog_name"
+	}
+	planObject, err := repository.Get(ctx, types.ServicePlanType,
+		query.ByField(query.EqualsOperator, key, selectorValue),
+		query.ByField(query.EqualsOperator, "service_offering_id", offeringID),
+	)
+	if err != nil {
+		return nil, util.HandleInstanceSharingError(util.ErrReferencedInstanceNotFound, "")
+	}
+	selectorPlan := planObject.(*types.ServicePlan)
+	return selectorPlan, nil
+}
+
+func matchLabels(instanceLabels types.Labels, labels []byte) (bool, error) {
+	var selectorLabels types.Labels
+	if err := util.BytesToObject(labels, &selectorLabels); err != nil {
+		return false, err
+	}
+	/*
+		todo: remove after review
+		should accept this label selector:
+			selectorLabels: {
+				"key": ["x"],
+				"region": ["us", "jp"]
+			}
+
+			instanceLabels: {
+				"key": ["x", "y"],
+				"region": ["us", "eu", "jp"]
+			}
+	*/
+
+	for selectorLabelKey, selectorLabelArray := range selectorLabels {
+		instanceLabelVal, exists := instanceLabels[selectorLabelKey]
+		if exists && instanceLabels != nil {
```

Comment:
change instanceLabelVal to instanceLabelArray it's confusing in contains(instanceLabelVal, selectorLabelVal)

Reply:
fixed

Reply:
you mean
if exists && len(instanceLabelArray) >0
right?